File: src/scripts/model/buildings/zones/zone.js
import {Building} from '../building.js';
import {PowerConsumer} from "../power/powerConsumer.js";

export const DevelopmentState = {
  abandoned: 'abandoned',
  developed: 'developed',
  underConstruction: 'under-construction',
  undeveloped: 'undeveloped',
};

export class Development {
  state = DevelopmentState.undeveloped;
  level = 0;
  abandonmentCounter = 0;
  constructionCounter = 0;
  maxLevel = 3;
}


/**
 * Represents a zoned building such as residential, commercial or industrial
 */
export class Zone extends Building {

  development = new Development();

  constructor(tile, type) {
    super(tile, type, new PowerConsumer(10), true)
  }

  /**
   * Returns an HTML representation of this object
   * @returns {string}
   */
  toHTML() {
    let html = super.toHTML() +  `
        <span class="info-label">State </span>
        <span class="info-value">${this.development.state}</span>
        <br>
        <span class="info-label">Level </span>
        <span class="info-value">${this.development.level}</span>
        <br>`;
    return html;
  }
}

File: Simulation/Model/Buildings/Zones/Zone.cs
namespace Simulation.Model.Buildings.Zones;

public static class DevelopmentState
{
    public const string Abandoned        = "abandoned";
    public const string Developed        = "developed";
    public const string UnderConstruction = "under-construction";
    public const string Undeveloped      = "undeveloped";
}

public class Development
{
    public string State                { get; set; } = DevelopmentState.Undeveloped;
    public int    Level                { get; set; }
    public int    AbandonmentCounter   { get; set; }
    public int    ConstructionCounter  { get; set; }
    public int    MaxLevel             { get; set; } = 3;
}

public class Zone : Building
{
    public Development Development { get; } = new();

    public Zone(Tile tile, string type)
        : base(tile, type, new PowerConsumer(10), needsRoadAccess: true) { }
}
